frame/include/bli_misc_macro_defs.h

```diff
@@ -170,5 +170,7 @@ BLIS_INLINE void bli_toggle_bool( bool* b )
 #define BLIS_VA_END  (-1)
 
 
-#endif
+// Static assertion compatible with any version of C/C++
+#define bli_static_assert(cond) while(0){struct s {int bli_static_assert : !!(cond);};}
```

Comment:
LOL, what is *this*? My brain can't even parse that syntax. 😂

Reply:
`!!` = make a conditional into 0 (false) or 1 ( true ). `int whatever : n` this is a bitfield with n>0 bits. Since a bitfield with 0 bits is a compile-time error, then a false condition leads to a compiler error (static assertion failure). One can also try an create and array e.g. `int blis_static_assert[!!(cond)]`. The `while(0)` is for the reason that I've mentioned a coupled of times, where just bare braces could get mis-parsed.

Reply:
This is one of those things where I say "live and let live." Weird-ass macro approved! ✅

Reply:
I will say it is traditional to name the variable (in the struct) something like 'STATIC_ASSERTION_FAILED' so that people see it in the output and aren't super confused.

Reply:
I tried a dozen syntaxes to try to find a `static_assert` (ala C++11) which is compatible with as many C/C++ dialects as possible, and settled on this one.

The error message is fairly clear, I think.

Let me break down the syntax:
```
while(0) { }
```
This makes it never execute, while still enforcing syntax/semantic checks, and making the whole macro expansion a single compound statement which can be unambiguously embedded inside an `if` statement without braces, similar to the `do { ... } while(0)` paradigm.
```
struct s {int bli_static_assert : !!(cond);};
```
This defines a struct which has a `bli_static_assert` bit-field member whose size depends on `cond`, which must be a compile-time constant (otherwise another error will occur).

If `cond` is true, then `!!(cond)` is `1`, and the bit-field is well-formed.

If `cond` is false, then `!!(cond)` is `0`, and since the bit-field is _named_, it is ill-formed, which causes a compilation error similar to this:
```
foo.c:1:56: error: zero width for bit-field ‘bli_static_assert’
    1 | #define bli_static_assert(cond) while(0){struct s {int bli_static_assert : !!(cond);};}
      |                                                        ^~~~~~~~~~~~~~~~~
foo.c:5:1: note: in expansion of macro ‘bli_static_assert’
    5 | bli_static_assert(0);
      | ^~~~~~~~~~~~~~~~~
```

It also points to the `bli_static_assert()` macro call where it failed.

The bit-field name can be changed to whatever you want it to be. I prefer it is not uppercase unless the macro name itself is uppercase. Since C++11 and C23 have `static_assert` (C11 has `_Static_assert`), I don't think it needs to be all-uppercase.

Reply:
Note: `int blis_static_assert[!!(cond)]` wouldn't work on compilers which support variable-length arrays, because it would allow `cond` to _not_ be a compile-time constant, and even if it was a compile-time constant known to be `0`, some compilers (not many) might even support 0-sized arrays as an extension. Bit fields are better than array sizes in enforcing compile-time constancy and non-zero sizes.

Reply:
Re the name I meant `struct s { int STATIC_ASSERT_FAILED : !!(cond)};`, `blis_static_assert` is good.

Reply:
> Re the name I meant `struct s { int STATIC_ASSERT_FAILED : !!(cond)};`, `blis_static_assert` is good.

I'm fine with that. I'll edit it.